avocado/core/runners/avocado_instrumented.py

```diff
@@ -32,6 +32,12 @@ class and method names should be separated by a ":".  One
     name = 'avocado-instrumented'
     description = 'Runner for Avocado INSTRUMENTED tests'
 
+    CONFIGURATION_USED = ['run.test_parameters',
+                          'datadir.paths.cache_dirs',
+                          'core.show',
+                          'job.output.loglevel',
+                          'job.run.store_logging_stream']
+
```

Comment:
By looking at the commit log, and the variable name I had the impression this would be used so runners could expose what namespaces they are using. However I could not find any of those keys inside of this runner code.

Could you help me to understand what I'm missing here, please?

Reply:
The `avocado-instrumented` runner is a bit harder to follow indeed, because it's execution does a lot of compatibility setup to run `avocado/core/test.py` (which by itself is not as decoupled as it should be be from `avocado/core/job.py`).

So, for instance, this is one of the indirect usages:

https://github.com/avocado-framework/avocado/blob/3599d2f98b9b93dfbc0db1950f2775ff41c72190/avocado/core/runners/avocado_instrumented.py#L80

Which in turn access one of those explicitly:

https://github.com/avocado-framework/avocado/blob/3599d2f98b9b93dfbc0db1950f2775ff41c72190/avocado/core/runners/utils/messages.py#L207

Reply:
Sounds to me that we keep introducing component dependencies, increasing our coupling levels.

IIUC: 1. the runner is saying it needs a namespace X; 2. the, avocado core is passing namespace X to runner; and 3; the runner is passing back the namespace X to a core library because it is not needed by the runner itself, but instead, by the core utility library.

Even if this understanding is wrong some how, seems to me we need a "remote config API": Let's say that we have a "remote config API" with just one read/get endpoint (we don't need to introduce an entire "etcd" here), something like: `remote_config.get(namespace)`. 

So, instead of passing those configs in an "one to one approach", what do you think of having a central place to ask for them using the network?

(I might be still missing a few things here)

Reply:
Sorry if I gave the wrong impression. I understand that is already present. The "we keep introducing component dependencies" it was more like a generic feeling, not specific to this PR (by re-reading now, I understand that it was not clear)

> One point to address with such proposal is the use case of running tests on disconnected environments, which is one of the original goals of the nrunner architecture.

But today the status server is collecting test results already, right?

Reply:
> Sorry if I gave the wrong impression. I understand that is already present. The "we keep introducing component dependencies" it was more like a generic feeling, not specific to this PR (by re-reading now, I understand that it was not clear)
> 

Sure, NP! Thanks for clarifying.

> > One point to address with such proposal is the use case of running tests on disconnected environments, which is one of the original goals of the nrunner architecture.
> 
> But today the status server is collecting test results already, right?

Yes, and the status server can continue to be the interface to inject test results into a job, even for tests running in disconnected environments. 

Right now, if spawners were to ask a runner running a task to *not* send results to a status server on the network (that is, omit the `-s` option), the runner will output the same information, but on its standard output.

For a disconnected operation, which we are missing Today, the spawner could collect that output and feed it into the results server, which, differently from the runner running the task itself, should always have the status server accessible.

I image that the role spawners play Today could be extended to ask what configurations are relevant and pass them along to runners.  My only problem with that is requiring that the spawners somehow have prior knowledge of runner types and what they need.  Because of that, IMO, the source of truth with regards to configuration need to be with the runners themselves, so that a true "plug and play" behavior is possible.

But I may be missing a way to consolidate that information outside of the runners.

Reply:
> Yes, and the status server can continue to be the interface to inject test results into a job, even for tests running in disconnected environments.

Let me play the Devil's advocate one more time ... :) Please, refresh my terrible memory: what are the use cases where the runners will not have connection (socket/network) to the main avocado component? How bad it would be requiring network on those runners?

>  I image that the role spawners play Today could be extended to ask what configurations are relevant and pass them along to runners. My only problem with that is requiring that the spawners somehow have prior knowledge of runner types and what they need. Because of that, IMO, the source of truth with regards to configuration need to be with the runners themselves, so that a true "plug and play" behavior is possible.

Right. I see and agree that the source of truth should be the runners if we take this approach. However, my "problem" with this design is that we are always passing values to very specific components, increasing coupling between them. And in this case we are passing in both directions. It would be nice to have an one-to-many (or even many-to-many) message bus or API, as [discussed](https://github.com/avocado-framework/avocado/discussions/4316) to loose this coupling. But indeed, today, I can't see this working in a fully network isolated runner. I guess my first question it will help to better evaluate this.

Reply:
> > Yes, and the status server can continue to be the interface to inject test results into a job, even for tests running in disconnected environments.
> 
> Let me play the Devil's advocate one more time ... :) Please, refresh my terrible memory: what are the use cases where the runners will not have connection (socket/network) to the main avocado component? How bad it would be requiring network on those runners?
> 

One of them, quite common with kernel, virtualization and similar fields, is to run a test within a minimal environment such as an initrd.  I can look for some references, but being able to check if "x works" in a kernel or in virtualized hardware without a lot of userspace having to run (or having run) would be very welcome to those fields.

The goal was recorded [here](https://avocado-framework.readthedocs.io/en/96.0/guides/contributor/chapters/runners.html):

_"Disconnected test execution, so that results can be saved to a device and collected by the runner."_ 

Although the intended meaning of "runner" in that sentence is not the `avocado-runner-*` but the "avocado test runner main application".

> > I image that the role spawners play Today could be extended to ask what configurations are relevant and pass them along to runners. My only problem with that is requiring that the spawners somehow have prior knowledge of runner types and what they need. Because of that, IMO, the source of truth with regards to configuration need to be with the runners themselves, so that a true "plug and play" behavior is possible.
> 
> Right. I see and agree that the source of truth should be the runners if we take this approach. However, my "problem" with this design is that we are always passing values to very specific components, increasing coupling between them. And in this case we are passing in both directions. It would be nice to have an one-to-many (or even many-to-many) message bus or API, as [discussed](https://github.com/avocado-framework/avocado/discussions/4316) to loose this coupling. But indeed, today, I can't see this working in a fully network isolated runner. I guess my first question it will help to better evaluate this.

I'm sure that the "passing values to very specific components" can be improved, and this is an attempt to do so.  For instance, look at [this](https://github.com/avocado-framework/avocado/pull/5313#issuecomment-1097991857) discussion.

Also, I think there is still value in the pub/sub approach.  Even with the disconnected runners approach it would be a suitable architecture, provided that the connectivity only goes up to the point of the component preparing/starting the execution and fetching the results.

Reply:
> > 
> > Let me play the Devil's advocate one more time ... :) Please, refresh my terrible memory: what are the use cases where the runners will not have connection (socket/network) to the main avocado component? How bad it would be requiring network on those runners?
> 
> One of them, quite common with kernel, virtualization and similar fields, is to run a test within a minimal environment such as an initrd. I can look for some references, but being able to check if "x works" in a kernel or in virtualized hardware without a lot of userspace having to run (or having run) would be very welcome to those fields.

Please let me know when you find the references, because by looking at your example, I'm still trying to understand why a socket would be a "a lot of userspace having to run". I do understand the initrd and the kernel being tested should be as minimal as possible, but iiuc the runner that is starting this would be in a different layer from what is being tested.  (sorry for the insistence, I'm not doubting, just pushing to us properly document, why we are investing so much to "preserve this feature", that is not supported currently. For the future me/others.)

> The goal was recorded [here](https://avocado-framework.readthedocs.io/en/96.0/guides/contributor/chapters/runners.html):
> 
> _"Disconnected test execution, so that results can be saved to a device and collected by the runner."_
> 
> Although the intended meaning of "runner" in that sentence is not the `avocado-runner-*` but the "avocado test runner main application".

Yes, I remember the "goal" but I was missing the "use case".

> I'm sure that the "passing values to very specific components" can be improved, and this is an attempt to do so. For instance, look at [this](https://github.com/avocado-framework/avocado/pull/5313#issuecomment-1097991857) discussion.

Perhaps we are not talking about the same components here. :/ With this we are for sure "improving" the amount of data to be passed. But still, we are passing data in an 1to1 approach: (ie: spawners send messages to runners directly; runners send messages to status server directly).

> Also, I think there is still value in the pub/sub approach. Even with the disconnected runners approach it would be a suitable architecture, provided that the connectivity only goes up to the point of the component preparing/starting the execution and fetching the results.

Having such pub/sub bus being used only up to the "dispatcher" machine, might be too much. I think the real advantage would be when using it on multiple remote components to avoid most of our "serialization/de serialization" steps. But perhaps this would be the first step to prove its value for a future "infestation".

Btw, here I was not proposing a PUB/SUB bus, just a simple remote call would make me happy. :) But yes... I do have this special use case to support that is not supported today.

Reply:
> > > Let me play the Devil's advocate one more time ... :) Please, refresh my terrible memory: what are the use cases where the runners will not have connection (socket/network) to the main avocado component? How bad it would be requiring network on those runners?
> > 
> > 
> > One of them, quite common with kernel, virtualization and similar fields, is to run a test within a minimal environment such as an initrd. I can look for some references, but being able to check if "x works" in a kernel or in virtualized hardware without a lot of userspace having to run (or having run) would be very welcome to those fields.
> 
> Please let me know when you find the references, because by looking at your example, I'm still trying to understand why a socket would be a "a lot of userspace having to run". I do understand the initrd and the kernel being tested should be as minimal as possible, but iiuc the runner that is starting this would be in a different layer from what is being tested. (sorry for the insistence, I'm not doubting, just pushing to us properly document, why we are investing so much to "preserve this feature", that is not supported currently. For the future me/others.)

A socket implies networking, not only at a kernel level, but with configuration usually performed by System-V or systemd-like environments.  Of course "a lot" is relative, but if I'm putting the hat of a kernel or virtualization developer, that could be a lot indeed.

When it comes to "the runner that is starting this would be in a different layer from what is being tested", that is a misunderstanding.  To be able to run an executable `check-device` under a development kernel, or run `check-real-time-latency-spikes` in a non-networked environment with less interruptions, the `avocado-runner-exec-test` would be in the same layer as the `check-device` or `check-real-time-latency-spikes`.

> 
> > The goal was recorded [here](https://avocado-framework.readthedocs.io/en/96.0/guides/contributor/chapters/runners.html):
> > _"Disconnected test execution, so that results can be saved to a device and collected by the runner."_
> > Although the intended meaning of "runner" in that sentence is not the `avocado-runner-*` but the "avocado test runner main application".
> 
> Yes, I remember the "goal" but I was missing the "use case".
> 
> > I'm sure that the "passing values to very specific components" can be improved, and this is an attempt to do so. For instance, look at [this](https://github.com/avocado-framework/avocado/pull/5313#issuecomment-1097991857) discussion.
> 
> Perhaps we are not talking about the same components here. :/ With this we are for sure "improving" the amount of data to be passed. But still, we are passing data in an 1to1 approach: (ie: spawners send messages to runners directly; runners send messages to status server directly).
> 

Right.

> > Also, I think there is still value in the pub/sub approach. Even with the disconnected runners approach it would be a suitable architecture, provided that the connectivity only goes up to the point of the component preparing/starting the execution and fetching the results.
> 
> Having such pub/sub bus being used only up to the "dispatcher" machine, might be too much. I think the real advantage would be when using it on multiple remote components to avoid most of our "serialization/de serialization" steps. But perhaps this would be the first step to prove its value for a future "infestation".
> 
> Btw, here I was not proposing a PUB/SUB bus, just a simple remote call would make me happy. :) But yes... I do have this special use case to support that is not supported today.

Sure, those are good questions to ask!